### PR TITLE
API for fetching current total number of built blocks.

### PR DESCRIPTION
**Pull request by mtrofin**

### CODE REVIEW

File: core/lib/instrument.c [hunk 1]
```diff
@@ -7595,6 +7595,13 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
     return true;
 }
 
+DR_API
+uint64
+dr_stats_get_built_blocks_count() {
```
Comment (derekbruening) [inline core/lib/instrument.c]: style violation: { on own line

File: core/lib/instrument.c [hunk 2]
```diff
@@ -7595,6 +7595,13 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
...
+  return GLOBAL_STAT(num_bbs);
```
Comment (derekbruening) [inline core/lib/instrument.c]: style: indentation

File: core/lib/instrument_api.h [hunk 3]
```diff
@@ -6201,6 +6201,13 @@ DR_API
 bool
 dr_prepopulate_cache(app_pc *tags, size_t tags_count);
 
+DR_API
+/**
+ * Get the number of blocks built so far, globally. The API is not thread-safe.
```
Comment (derekbruening) [inline core/lib/instrument_api.h]: "blocks" should be clearer: "basic blocks".  The comment could also explain that this is the total number ever built and includes duplicates if there were deletions for consistency (i.e., app code mod) or capacity reasons or thread-private caches.

File: core/lib/instrument_api.h [hunk 4]
```diff
@@ -6201,6 +6201,13 @@ DR_API
...
+ */
+uint64
+dr_stats_get_built_blocks_count();
```
Comment (derekbruening) [inline core/lib/instrument_api.h]: Hmm, if the plan is to provide a set of statistics that might grow to a dozen or more, maybe it is better to return a pointer to a struct that we can extend, making it much easier to add new statistics without having to add a whole new API function?

Comment (mtrofin) [reply core/lib/instrument_api.h]: Done.

File: core/lib/instrument_api.h (hunk 3 above)
Comment (mtrofin) [reply core/lib/instrument_api.h]: Done.

File: core/lib/instrument.c (hunk 2 above)
Comment (mtrofin) [reply core/lib/instrument.c]: Done

File: core/lib/instrument.c (hunk 1 above)
Comment (mtrofin) [reply core/lib/instrument.c]: Done

**Comment by mtrofin** [comment]
This addresses issue #2928

File: core/lib/instrument_api.h [hunk 5]
```diff
@@ -6201,12 +6201,28 @@ DR_API
...
 dr_prepopulate_cache(app_pc *tags, size_t tags_count);
 
+/* DR_API EXPORT BEGIN */
+/** Used with dr_get_stats().*/
+typedef struct _dr_stats_t {
+    /** The size of this structure. Set this to sizeof(dr_stats_t). */
+    size_t size;
+    /** The total number of basic blocks ever built so far, globally. This
+     *  includes duplicates and blocks that were deleted for consistency
+     *  or capacity reasons or thread-private caches.
```
Comment (derekbruening) [inline core/lib/instrument_api.h]: for

File: core/lib/instrument.c [hunk 6]
```diff
@@ -7596,9 +7596,17 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
 }
 
 DR_API
-uint64
-dr_stats_get_built_blocks_count() {
-  return GLOBAL_STAT(num_bbs);
+bool
+dr_get_stats(dr_stats_t* drstats)
```
Comment (derekbruening) [inline core/lib/instrument.c]: style: t *d

File: core/lib/instrument.c [hunk 7]
```diff
@@ -7596,9 +7596,17 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
 }
```
Comment (derekbruening) [inline core/lib/instrument.c]: Please add an entry to the changelist of new features at https://github.com/DynamoRIO/dynamorio/blob/master/api/docs/release.dox#L226

File: core/lib/instrument.c [hunk 8]
```diff
@@ -7596,9 +7596,17 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
...
 
 DR_API
-uint64
-dr_stats_get_built_blocks_count() {
-  return GLOBAL_STAT(num_bbs);
+bool
+dr_get_stats(dr_stats_t* drstats)
+{
+    if (!GLOBAL_STATS_ON()) return false;
```
Comment (derekbruening) [inline core/lib/instrument.c]: style: body on own line

File: core/lib/instrument.c [hunk 9]
```diff
@@ -7595,6 +7595,22 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
     return true;
 }
 
+DR_API
+bool
+dr_get_stats(dr_stats_t* drstats)
```
Comment (derekbruening) [inline core/lib/instrument.c]: style: t *d

File: core/lib/instrument.c [hunk 10]
```diff
@@ -7595,6 +7595,22 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
...
+{
+    if (!GLOBAL_STATS_ON())
+        return false;
+    ASSERT(drstats != NULL);
```
Comment (derekbruening) [inline core/lib/instrument.c]: Use CLIENT_ASSERT with a string message

File: core/lib/instrument.c [hunk 11]
```diff
@@ -7595,6 +7595,22 @@ dr_prepopulate_cache(app_pc *tags, size_t tags_count)
...
+    /* We are at V1 of the structure, and we can't return less than the one
+     * field. We need to remove this assert when we add more fields.
+     */
+    ASSERT(drstats->size >= sizeof(dr_stats_t));
```
Comment (derekbruening) [inline core/lib/instrument.c]: Use CLIENT_ASSERT with a string message

File: core/lib/instrument.c (hunk 9 above)
Comment (mtrofin) [reply core/lib/instrument.c]: done

File: core/lib/instrument.c (hunk 7 above)
Comment (mtrofin) [reply core/lib/instrument.c]: done

File: core/lib/instrument.c (hunk 10 above)
Comment (mtrofin) [reply core/lib/instrument.c]: Done

File: core/lib/instrument.c (hunk 11 above)
Comment (mtrofin) [reply core/lib/instrument.c]: Done

File: core/lib/instrument_api.h (hunk 5 above)
Comment (mtrofin) [reply core/lib/instrument_api.h]: Not sure I follow :)

Comment (derekbruening) [reply core/lib/instrument_api.h]: s/or/for/

Comment (derekbruening) [reply core/lib/instrument_api.h]: Wait n/m, misread it.

**Comment by derekbruening** [comment]
run aarch32 tests

**Comment by derekbruening** [comment]
run aarch64 tests

**Pull request merged**